### PR TITLE
updated upload plugin

### PR DESCRIPTION
**Pull request by giovannialbero1992**
# Description

I implemented during the past 2 weeks this feature and unfortunately I noticed that was already there in the project the feature but anyway, considering that IMHO the business logic should be decoupled by the router I'm proposing my solution here.

Of course feel free to discard it if you prefer to keep the solution at it is.

Thanks again for everything
See you

Related to issue #(issue)

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas

### CODE REVIEW

**Comment by pieroit** [comment]
Hi @AlboCode thanks for this

As you already noticed this part was already merged, so it pains my hearth that we've done the same work in parallel.
Next time we'll be mor eattentive to write explicit issues and have an explicit asseegnee.

Can we save somthing from this PR?

**Comment by giovannialbero1992** [comment]
Don't worry @pieroit, I'd like to understand if we can keep more decoupled the application layer by the infrastructure.
So to have a code more reliable and maintainable.
I'll try to refactor my code in order to give you a possible solution, saving first of all the tests that I added :)

**Comment by pieroit** [comment]
@giovannialbero1992 sounds good, having well separated classes is great to make the Cat stable and testable.

I see you are starting the unit tests which is a good idea (p.s. we are using pytest!)

Take a look here for the end2end testing of plugin upload:
https://github.com/cheshire-cat-ai/core/blob/main/core/tests/routes/test_routes_plugins.py#L32

As you can see the zip file is created "on fly" from the source code of a mock plugin.

Let me know for any question, we can also have a call in discord open mic to know each other :)

Thanks, your ideas and precision are appreciated

**Comment by giovannialbero1992** [comment]
Hi @pieroit, I just updated the PR if you wanna take a view :)

File: core/cat/routes/plugins.py
Comment (pieroit) [inline core/cat/routes/plugins.py:62]: Nice, agree

File: core/tests/infrastructure/test_zip.py
Comment (pieroit) [inline core/tests/infrastructure/test_zip.py:6]: No need for zip files in the repo. Please create a zip  of the `mock_plugin` via code, as done [here](https://github.com/cheshire-cat-ai/core/blob/f579b4badd942bb6862a055fcebb959db788d3c5/core/tests/routes/test_routes_plugins.py#L44)

File: core/cat/infrastructure/package.py
Comment (pieroit) [inline core/cat/infrastructure/package.py:21]: why not `shutil` ? Should be able to manage both zip and tar

Comment (pieroit) [reply core/cat/infrastructure/package.py:21]: (I'm asking ;))

File: core/cat/mad_hatter/mad_hatter.py
Comment (pieroit) [inline core/cat/mad_hatter/mad_hatter.py:29]: it is a good idea to put this in the mad hatter, thanks

File: core/cat/mad_hatter/mad_hatter.py
Comment (pieroit) [inline core/cat/mad_hatter/mad_hatter.py:45]: Yes!

**Comment by pieroit** [comment]
@giovannialbero1992 a few things to fix, you should find a review with several comments.
After these fixes, I'll merge and if you are available we can have a call to better refactor the whole `mad_hatter`
Thansk a lot :)

File: core/tests/mad_hatter/test_mad_hatter.py
Comment (pieroit) [inline core/tests/mad_hatter/test_mad_hatter.py:12]: Love that you mock a different plugin folder (I'll update to do it also [here](https://github.com/cheshire-cat-ai/core/blob/main/core/tests/routes/test_routes_plugins.py#L33)).
Question: we are using `pytest` for testing, here you use utilities from `unittest`. Can we do the mocks [as in here](https://github.com/cheshire-cat-ai/core/blob/05e97cc569128dbdf2580bf95fc65f8738169de5/core/tests/conftest.py#L33) or are we forced to combine the two libraries?

File: core/tests/mad_hatter/test_mad_hatter.py
Comment (pieroit) [inline core/tests/mad_hatter/test_mad_hatter.py:17]: Please no zip files in the repo, create the mock plugin zip at runtime as done [here](https://github.com/cheshire-cat-ai/core/blob/main/core/tests/routes/test_routes_plugins.py#L42) (using the Zip class maybe)

It is useful so we can commit on the [mock plugin](https://github.com/cheshire-cat-ai/core/tree/main/core/tests/mocks/mock_plugin) sources

File: core/cat/mad_hatter/mad_hatter.py
Comment (pieroit) [inline core/cat/mad_hatter/mad_hatter.py:43]: if we call bootstrap here, there is no need to call `find_plugin` above, because the `mad_hatter` makes the discovery at instantiation

**Review by pieroit — CHANGES_REQUESTED** [review]

**Comment by pieroit** [comment]
@giovannialbero1992 sorry for being picky, this PR is really important.
I left a review with a few additional requests.
Thanks a lot for your patience :)

**Comment by giovannialbero1992** [comment]
You are right @pieroit. The PR is upgraded :slightly_smiling_face:

**Comment by pieroit** [comment]
@giovannialbero1992 thanks.

A few things:
- I had to deactivate tests in which you call the CheshireCat constructor. That messes up badly things until we do a good refactor. Code is still there so it will be reactivated at the right time
- you touched on a few things for which there were no tests, so I added them ;)
- next time smaller PRs, unless we discussed things in an issue or in the Discord.

Great work thanks

**Pull request merged**
